ducc/lib/podman_store.go

```diff
@@ -0,0 +1,534 @@
+		return err
+	}
+	return nil
+}
+
+// Ingest the exploded rootfs in podman store.
+func (img *Image) IngestRootfsIntoPodmanStore(CVMFSRepo, subDirInsideRepo string, digestMap map[string]string) (err error) {
+	Log().WithFields(log.Fields{"action": "Ingesting layer rootfs into podman store for the image"}).Info(img.GetSimpleName())
+	manifest, err := img.GetManifest()
+	if err != nil {
+		LogE(err).Warn("Error in getting the image manifest")
+		return err
+	}
+	for _, layer := range manifest.Layers {
+		layerid := strings.Split(layer.Digest, ":")[1]
+		layerdir := digestMap[layer.Digest]
+
+		//symlinkPath will contain the rootfs of the corresponding layer in podman store.
+		symlinkPath := filepath.Join(rootPath, rootfsDir, layerdir, "diff")
+		targetPath := filepath.Join(subDirInsideRepo, layerid[:2], layerid, "layerfs")
+
+		if _, err := os.Stat(symlinkPath); os.IsNotExist(err) {
+			err = CreateSymlinkIntoCVMFS(CVMFSRepo, symlinkPath, targetPath)
+			if err != nil {
+				LogE(err).Error("Error in creating the symlink for the diff dir")
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Create the link dir and link file for exploded rootfs.
+func (img *Image) CreateLinkDir(CVMFSRepo, subDirInsideRepo string, digestMap, layerIdMap map[string]string) (err error) {
+	Log().WithFields(log.Fields{"action": "Creating Link files for layer rootfs for the image"}).Info(img.GetSimpleName())
+	manifest, err := img.GetManifest()
+	if err != nil {
+		LogE(err).Warn("Error in getting the image manifest")
+		return err
+	}
+	for _, layer := range manifest.Layers {
+		layerdir := digestMap[layer.Digest]
+		linkPath := filepath.Join(rootPath, rootfsDir, layerdir, "link")
+		if _, err := os.Stat(linkPath); os.IsNotExist(err) {
+			//generate the link id
+			lid, err := generateID(26)
+			if err != nil {
+				LogE(err).Error("Error generating file name for Link dir")
+				return err
+			}
+			layerIdMap[layer.Digest] = filepath.Join("l", lid)
+			err = writeDataToCvmfs(CVMFSRepo, linkPath, []byte(lid))
+			if err != nil {
+				LogE(err).Error("Error in writing link id to podman store")
+				return err
+			}
+
+			//Create link dir
+			symlinkPath := filepath.Join(rootPath, rootfsDir, "l", lid)
+			targetPath := filepath.Join(rootPath, rootfsDir, layerdir, "diff")
+
+			err = CreateSymlinkIntoCVMFS(CVMFSRepo, symlinkPath, targetPath)
+			if err != nil {
+				LogE(err).Error("Error in creating the symlink for the Link dir")
+				return err
+			}
+		} else {
+			data, err := ioutil.ReadFile(linkPath)
+			if err != nil {
+				LogE(err).Error("Error in reading link file")
+				return err
+			}
+			layerIdMap[layer.Digest] = string(data)
+		}
+	}
+	return nil
+}
+
+// Create the lower files for diff dirs to be used by podman.
+func (img *Image) CreateLowerFiles(CVMFSRepo string, digestMap, layerIdMap map[string]string) (err error) {
+	Log().WithFields(log.Fields{"action": "Creating Lower files for diff dirs in podman store"}).Info(img.GetSimpleName())
```

Comment:
Is it the lower _layers_? I think that would make more sense than calling it `lower files`

Reply:
No, actually they store the directory names of the lower layers. Podman names these files as "lower" , hence this name.

Reply:
I see, but then it's "lower directories" instead of "lower files", no?

Reply:
No, they are files which contain a string (name of the parent layers) as its content.  Podman names these files as lower files

Reply:
I see. Thanks!